src/countervalues/react.js

```diff
@@ -0,0 +1,457 @@
+// @flow
+import { BigNumber } from "bignumber.js";
+import React, {
+  createContext,
+  useMemo,
+  useContext,
+  useEffect,
+  useReducer,
+  useState,
+  useRef,
+  useCallback,
+} from "react";
+import type {
+  Account,
+  AccountLike,
+  PortfolioRange,
+  Currency,
+  CryptoCurrency,
+  TokenCurrency,
+} from "../types";
+import {
+  getBalanceHistoryWithCountervalue,
+  getPortfolio,
+  getCurrencyPortfolio,
+  getAssetsDistribution,
+} from "../portfolio";
+import {
+  getAccountCurrency,
+  flattenAccounts,
+  getAccountUnit,
+} from "../account";
+import {
+  initialState,
+  calculate,
+  calculateMany,
+  loadCountervalues,
+  exportCountervalues,
+  importCountervalues,
+} from "./logic";
+import type {
+  CounterValuesState,
+  CounterValuesStateRaw,
+  CountervaluesSettings,
+} from "./types";
+
+// Polling is the control object you get from the high level <PollingConsumer>{ polling => ...
+export type Polling = {
+  // completely wipe all countervalues
+  wipe: () => void,
+  // one shot poll function
+  // TODO: is there any usecases returning promise here?
+  // It's a bit tricky to return Promise with current impl
+  poll: () => void,
+  // start background polling
+  start: () => void,
+  // stop background polling
+  stop: () => void,
+  // true when the polling is in progress
+  pending: boolean,
+  // if the last polling failed, there will be an error
+  error: ?Error,
+};
+
+export type Props = {
+  children: React$Node,
+  userSettings: CountervaluesSettings,
+  // the time to wait before the first poll when app starts (allow things to render to not do all at boot time)
+  pollInitDelay?: number,
+  // the minimum time to wait before two automatic polls (then one that happen whatever network/appstate events)
+  autopollInterval?: number,
+  savedState?: CounterValuesStateRaw,
+};
+
+const CountervaluesPollingContext = createContext<Polling>({
+  wipe: () => {},
+  poll: () => {},
+  start: () => {},
+  stop: () => {},
+  pending: false,
+  error: null,
+});
+
+const CountervaluesContext = createContext<CounterValuesState>(initialState);
+
+export function Countervalues({
+  children,
+  userSettings,
+  pollInitDelay = 1 * 1000,
+  autopollInterval = 120 * 1000,
+  savedState,
+}: Props) {
+  const [{ state, pending, error }, dispatch] = useReducer<FetchState, Action>(
+    fetchReducer,
+    initialFetchState
+  );
+
+  // trigger poll but not every time poll callback is changed
+  const [triggerPoll, setTriggerPoll] = useState(false);
+
+  useEffect(() => {
+    if (pending || !triggerPoll) return;
+    dispatch({ type: "pending" });
+    loadCountervalues(state, userSettings).then(
+      (state) => {
```

Comment:
⚠️  there are race condition here if a second effect runs before the end of the first